### PR TITLE
Add CoveringIndexTrait and CoveringIndexConfigTrait

### PR DESCRIPTION
**Pull request by sezruby**
### What is the context for this pull request?
 - **Tracking Issue**: n/a
 - **Parent Issue**: #515
 - **Dependencies**: n/a

### What changes were proposed in this pull request?
CoveringIndex refactoring to support other covering index type - ZOrderCoveringIndex (#495)
Added CoveringIndexTrait & CoveringIndexConfigTrait to reduce duplicate code.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Refactoring change - tested by existing tests

### CODE REVIEW

File: src/main/scala/com/microsoft/hyperspace/index/covering/CoveringIndexTrait.scala [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+import com.microsoft.hyperspace.index._
+import com.microsoft.hyperspace.util.ResolverUtils.ResolvedColumn
+
+/**
+ * CoveringIndexTrait is a common interface for covering index type which contains a vertical
+ * slice of source data including indexedColumns and includedColumns.
+ */
+trait CoveringIndexTrait extends Index {
+  val schema: StructType
+  val includedColumns: Seq[String]
```
Comment (clee704) [inline src/main/scala/com/microsoft/hyperspace/index/covering/CoveringIndexTrait.scala:34]: nit: I would rather not use `val` in traits unless there is a specific reason.

Refs:
1. https://alvinalexander.com/scala/fp-book/def-vs-val-abstract-members-traits-classes-override/
2. https://stackoverflow.com/questions/19642053/when-to-use-val-or-def-in-scala-traits

File: src/main/scala/com/microsoft/hyperspace/index/covering/CoveringIndex.scala [hunk 2]
```diff
@@ -19,39 +19,41 @@ package com.microsoft.hyperspace.index.covering
...
 
-  override def referencedColumns: Seq[String] = indexedColumns ++ includedColumns
-
   override def withNewProperties(newProperties: Map[String, String]): CoveringIndex = {
     copy(properties = newProperties)
   }
 
-  override def statistics(extended: Boolean = false): Map[String, String] = {
-    simpleStatistics ++ (if (extended) extendedStatistics else Map.empty)
+  protected def copyIndex(
```
Comment (clee704) [inline src/main/scala/com/microsoft/hyperspace/index/covering/CoveringIndex.scala:49]: nit: `override`

File: src/main/scala/com/microsoft/hyperspace/index/covering/CoveringIndex.scala [hunk 3]
```diff
@@ -19,39 +19,41 @@ package com.microsoft.hyperspace.index.covering
...
+      indexedCols: Seq[String],
+      includedCols: Seq[String],
+      schema: StructType): CoveringIndex = {
+    copy(indexedColumns = indexedCols, includedColumns = includedCols, schema = schema)
   }
 
-  override def canHandleDeletedFiles: Boolean = hasLineageColumn
-
-  override def write(ctx: IndexerContext, indexData: DataFrame): Unit = {
+  protected def write(ctx: IndexerContext, indexData: DataFrame, mode: SaveMode): Unit = {
```
Comment (clee704) [inline src/main/scala/com/microsoft/hyperspace/index/covering/CoveringIndex.scala:56]: nit: `override`

File: src/main/scala/com/microsoft/hyperspace/index/covering/CoveringIndex.scala [hunk 4]
```diff
@@ -170,22 +84,21 @@ case class CoveringIndex(
...
-  def hasLineageColumn: Boolean = IndexUtils.hasLineageColumn(properties)
+  def bucketSpec: Option[BucketSpec] =
+    Some(
+      BucketSpec(
+        numBuckets = numBuckets,
+        bucketColumnNames = indexedColumns,
+        sortColumnNames = indexedColumns))
 
-  private def simpleStatistics: Map[String, String] = {
+  protected def simpleStatistics: Map[String, String] = {
```
Comment (clee704) [inline src/main/scala/com/microsoft/hyperspace/index/covering/CoveringIndex.scala:94]: nit: `override`

File: src/main/scala/com/microsoft/hyperspace/index/covering/CoveringIndex.scala [hunk 5]
```diff
@@ -170,22 +84,21 @@ case class CoveringIndex(
...
     Map(
       "includedColumns" -> includedColumns.mkString(", "),
       "numBuckets" -> numBuckets.toString,
       "schema" -> schema.json)
   }
 
-  private def extendedStatistics: Map[String, String] = {
+  protected def extendedStatistics: Map[String, String] = {
```
Comment (clee704) [inline src/main/scala/com/microsoft/hyperspace/index/covering/CoveringIndex.scala:101]: nit: `override`

**Review by clee704 — APPROVED** [review]

File: src/main/scala/com/microsoft/hyperspace/index/covering/CoveringIndexTrait.scala (hunk 1 above)
Comment (sezruby) [reply src/main/scala/com/microsoft/hyperspace/index/covering/CoveringIndexTrait.scala:34]: Thanks! I'll address your comments in the next PR.

**Pull request merged**